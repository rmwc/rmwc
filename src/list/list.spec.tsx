import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  List,
  ListItem,
  ListItemPrimaryText,
  ListItemGraphic,
  ListItemMeta,
  SimpleListItem,
  CollapsibleList
} from './';

describe('List', () => {
  it('renders', () => {
    const { asFragment } = render(
      <List>
        <ListItem ripple>
          <ListItemGraphic />
          <ListItemPrimaryText>Cookies</ListItemPrimaryText>
          <ListItemMeta />
        </ListItem>
        <ListItem ripple={false}>
          <ListItemGraphic />
          <ListItemPrimaryText>Cookies</ListItemPrimaryText>
          <ListItemMeta />
        </ListItem>
      </List>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles onAction', async () => {
    let clickedIndex: any;

    render(
      <List onAction={(evt) => (clickedIndex = evt.detail)}>
        <ListItem>
          <ListItemPrimaryText>Cookies</ListItemPrimaryText>
        </ListItem>
        <ListItem>
          <ListItemPrimaryText>Pancakes</ListItemPrimaryText>
        </ListItem>
      </List>
    );

    userEvent.click(screen.getByText('Pancakes'));

    await waitFor(() => expect(clickedIndex).toEqual({ index: 1 }));

    userEvent.click(screen.getByText('Cookies'));

    await waitFor(() => expect(clickedIndex).toEqual({ index: 0 }));
  });

  it('SimpleListItem renders', () => {
    const { asFragment } = render(
      <List>
        <SimpleListItem graphic="star_border" text="Cookies" />
        <SimpleListItem
          graphic="star_border"
          text="Cookies"
          secondaryText="Chocolate chip"
        />
        <SimpleListItem
          graphic="star_border"
          text="Cookies"
          secondaryText="Chocolate chip"
          meta="info"
        />
      </List>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('SimpleListItem can have children', () => {
    render(
      <SimpleListItem
        graphic="star_border"
        text="Cookies"
        secondaryText="Chocolate chip"
        meta="info"
      >
        <aside>Test</aside>
      </SimpleListItem>
    );
    expect(screen.getByText('Test')).toBeInTheDocument();
  });

  it('can have custom classnames', () => {
    [List, ListItem, ListItemPrimaryText].forEach(
      (Component: React.ComponentType<any>) => {
        const { container } = render(
          <Component className={'my-custom-classname'} />
        );
        expect(container.firstChild).toHaveClass('my-custom-classname');
      }
    );
  });

  it('can be activated', () => {
    const { container } = render(<ListItem activated />);
    expect(container.firstChild).toHaveClass(
      'mdc-deprecated-list-item--activated'
    );
  });

  it('can be selected', () => {
    const { container } = render(<ListItem selected />);
    expect(container.firstChild).toHaveClass(
      'mdc-deprecated-list-item--selected'
    );
  });
});

describe('Collapsible List', () => {
  it('renders', () => {
    const { asFragment } = render(
      <List>
        <ListItem>One</ListItem>
        <CollapsibleList handle={<ListItem>Handle</ListItem>}>
          <ListItem>Two</ListItem>
        </CollapsibleList>
      </List>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles lifecycle', () => {
    render(
      <List>
        <ListItem>One</ListItem>
        <CollapsibleList
          defaultOpen
          handle={<ListItem className="handle">Handle</ListItem>}
        >
          <ListItem>Two</ListItem>
        </CollapsibleList>
      </List>
    );

    userEvent.click(screen.getByText('One'));

    expect(screen.getByText('Handle')).toBeInTheDocument();
  });

  it('handles events: enter', () => {
    render(
      <List>
        <ListItem>One</ListItem>
        <CollapsibleList
          defaultOpen
          handle={<ListItem className="handle">Handle</ListItem>}
        >
          <ListItem>Two</ListItem>
        </CollapsibleList>
      </List>
    );

    userEvent.type(screen.getByText('One'), '{enter}');
    expect(screen.getByText('Handle')).toBeInTheDocument();
  });
});
