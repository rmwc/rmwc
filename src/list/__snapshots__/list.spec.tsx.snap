// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible List renders 1`] = `
<DocumentFragment>
  <ul
    class="mdc-list"
  >
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      One
    </li>
    <div
      class="rmwc-collapsible-list"
    >
      <div
        class="rmwc-collapsible-list__handle"
      >
        <li
          class="mdc-list-item"
          tabindex="0"
        >
          <div
            class="mdc-list-item__ripple "
          />
          Handle
        </li>
      </div>
      <div
        class="rmwc-collapsible-list__children"
        style="max-height: 0px;"
      >
        <div
          class="rmwc-collapsible-list__children-inner"
        >
          <li
            class="mdc-list-item"
            tabindex="0"
          >
            <div
              class="mdc-list-item__ripple "
            />
            Two
          </li>
        </div>
      </div>
    </div>
  </ul>
</DocumentFragment>
`;

exports[`List SimpleListItem renders 1`] = `
<DocumentFragment>
  <ul
    class="mdc-list"
  >
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      <div
        class="mdc-list-item__ripple "
      />
      <i
        aria-hidden="true"
        class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
      >
        star_border
      </i>
      Cookies
    </li>
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      <div
        class="mdc-list-item__ripple "
      />
      <i
        aria-hidden="true"
        class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
      >
        star_border
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          Cookies
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          Chocolate chip
        </span>
      </span>
    </li>
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      <div
        class="mdc-list-item__ripple "
      />
      <i
        aria-hidden="true"
        class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
      >
        star_border
      </i>
      <span
        class="mdc-list-item__text"
      >
        <span
          class="mdc-list-item__primary-text"
        >
          Cookies
        </span>
        <span
          class="mdc-list-item__secondary-text"
        >
          Chocolate chip
        </span>
      </span>
      <div
        class="mdc-list-item__meta"
      >
        info
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`List renders 1`] = `
<DocumentFragment>
  <ul
    class="mdc-list"
  >
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      <i
        aria-hidden="true"
        class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
      />
      <span
        class="mdc-list-item__primary-text"
      >
        Cookies
      </span>
      <div
        class="mdc-list-item__meta"
      />
    </li>
    <li
      class="mdc-list-item"
      tabindex="0"
    >
      <div
        class="mdc-list-item__ripple "
      />
      <i
        aria-hidden="true"
        class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__graphic"
      />
      <span
        class="mdc-list-item__primary-text"
      >
        Cookies
      </span>
      <div
        class="mdc-list-item__meta"
      />
    </li>
  </ul>
</DocumentFragment>
`;
